### ISSUE TITLE
False positive: core.NullDereference on a const, non-null pointer member

### ISSUE THREAD

**Issue by GuB-42** [llvm/llvm-project#105560]
With the following code

```C++
struct A {
	void test(bool doSet, int v);
	int *tPtr = nullptr;
	int val = 0;
	int * const nonNullPtr = &val;
};

#if 1

void A::test(bool doSet, int v) {
	if (doSet) tPtr = nonNullPtr;
	*(tPtr ? tPtr : nonNullPtr) = v;
}

#else

int main() {
	A a;
	a.test(false, 42);
	return 0;
}

#endif
```

With options 

```
--analyze --analyzer-output text -Xanalyzer -analyzer-checker=core.NullDereference -Xclang -analyzer-checker=core,debug.ExprInspection
```

I get

```
<source>:12:30: warning: Dereference of null pointer [core.NullDereference]
   12 |         *(tPtr ? tPtr : nonNullPtr) = v;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
<source>:11:6: note: Assuming 'doSet' is true
   11 |         if (doSet) tPtr = nonNullPtr;
      |             ^~~~~
<source>:11:2: note: Taking true branch
   11 |         if (doSet) tPtr = nonNullPtr;
      |         ^
<source>:12:4: note: Assuming field 'tPtr' is null
   12 |         *(tPtr ? tPtr : nonNullPtr) = v;
      |           ^~~~
<source>:12:4: note: '?' condition is false
<source>:12:30: note: Dereference of null pointer
   12 |         *(tPtr ? tPtr : nonNullPtr) = v;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
1 warning generated.
```

https://godbolt.org/z/vE7hWbWjh

There is no way for nonNullPtr to be NULL besides memory corruption , it is const and initialized to a valid target.

If A::test() is defined and used in the same translation unit, the warning disappears. It can be seen by removing the #else directive in the example code. For building a practical example that still shows the warning, have the A::test() definition and main() in different files.

Tested on various clang versions, including 11, 16, 18 and trunk. Also with clang-tidy.
